cli/Renderer.ts

```diff
@@ -0,0 +1,95 @@
+"use strict";
+
+import { Map, Iterable, OrderedSet } from "immutable";
+import { TopLevels } from "./Type";
+import { Named, Namespace, assignNames } from "./Naming";
+import {
+    Source,
+    Sourcelike,
+    NewlineSource,
+    sourcelikeToSource,
+    serializeSource,
+    newline
+} from "./Source";
+
+export type RenderResult = { source: Source; names: Map<Named, string> };
+
+export abstract class Renderer {
+    protected readonly topLevels: TopLevels;
+    private _names: Map<Named, string> | undefined;
+
+    private _lastNewline?: NewlineSource;
+    private _emitted: Sourcelike[];
+
+    constructor(topLevels: TopLevels) {
+        this.topLevels = topLevels;
+        this._emitted = [];
+    }
+
+    emitNewline(): void {
+        const nl = newline();
+        this._emitted.push(nl);
+        this._lastNewline = nl;
+    }
+
+    emitLine(line: Sourcelike): void {
```

Comment:
You can make this variadic:

```typescript
emitLine(...lineParts: Sourcelike[]) void {
   lineParts.forEach(this._emitted.push);
```

We pass array literals into this pretty often, so we could get a nice readability boost.